x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/Attribute.java

```diff
@@ -32,6 +32,55 @@
  * The rest are not as they are not part of the projection and thus are not part of the derived table.
  */
 public abstract class Attribute extends NamedExpression {
```

Comment:
The attribute hierarchy is getting scarier for me.
This is probably outside of this change
I wonder if it is worth converting it to a flat record with a type field and a number of factories?

Something like:
```
Attribute.EMPTY
Attribute.unresolved(..)
Attribute.unsupported(..)
Attribute.metadata(..)
Attribute.field(..)
...
```

this will have the super set of all fields needed for everything but will be much easier to check equality and type (opposed to instanceof). Updating serialization accordingly might be a chllenge.

Reply:
Hm, maybe that's possible, but will mess with many `instanceOf` checks that we rely on. Let's consider this for another time.